torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -346,16 +371,19 @@ def _param_to_rank(self) -> Dict[torch.Tensor, int]:
         return self._param_to_rank_cache
 
     @property
-    def _param_to_index(self) -> Dict[int, int]:
+    def _param_id_to_index(self) -> Dict[int, int]:
         r"""
-        Hash table mapping parameters to their indices in the global optimizer
-        scheme.
+        Hash table mapping parameter IDs to their indices in the global
+        optimizer state.
+
+        NOTE: This assumes that the global optimizer state's indexing (in
+        ``state_dict``) follows a linear ordering over the parameter groups.
         """
-        if len(self._param_to_index_cache) == 0:
-            self._param_to_index_cache = {
+        if len(self._param_id_to_index_cache) == 0:
+            self._param_id_to_index_cache = {
                 id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))
```

Comment:
(seconding that personally, but soft preference)

Reply:
I examined the hashing for both `Tensor` and `Parameter` and neither seem to use `id()` by default.
```
>>> model = torch.nn.Linear(1, 1)
>>> t = model.weight.data  # t: torch.Tensor
>>> d = dict()
>>> d[t] = 1
>>> d[id(t)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 139809367343568
>>> p = model.weight  # p: torch.nn.parameter.Parameter
>>> d = dict()
>>> d[p] = 1
>>> d[id(p)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 139809367327664
```
Let me know if I tested this incorrectly.

Reply:
Thanks for checking this. My mistake.

Reply:
Hold on, I am not sure the test is correct. We are passing a `Tensor: int` pair, but checking an `int: int` pair? Isn't that suppose to fail? Below is Tensor'e hash function:

https://github.com/pytorch/pytorch/blob/8dd1dc89cbe1154640aa058b1eb09682144cfedd/torch/_tensor.py#L603-L606

Reply:
Ah, I misunderstood. Yes, I believe we can switch back to `_param_to_index` and make it of type `dict[torch.Tensor, int]`. Then, instead of the key being `id(p)`, it can directly be `p`.